### ISSUE TITLE
Index Lifecycle Allocation Routed Step failing with NullPtrException

### ISSUE THREAD

**Issue by DisasteR** [elastic/elasticsearch#63877]
**Elasticsearch version** (`bin/elasticsearch --version`):

```
Version: 6.8.12, Build: default/deb/7a15d2a/2020-08-12T07:27:20.804867Z, JVM: 1.8.0_252
```

**Plugins installed**: []

**JVM version** (`java -version`):

```
openjdk version "1.8.0_252"
OpenJDK Runtime Environment (build 1.8.0_252-8u252-b09-1~18.04-b09)
OpenJDK 64-Bit Server VM (build 25.252-b09, mixed mode)
```

**OS version** (`uname -a` if on a Unix-like system):

```
Linux esl-master-01 4.15.0-55-generic #60-Ubuntu SMP Tue Jul 2 18:22:20 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Description of the problem including expected versus actual behavior**:

Following https://discuss.elastic.co/t/index-lifecycle-allocation-routed-step-failing-with-nullptrexception/252331 i open a issue here

We are running an ElasticSearch 6.8.12 cluster.

We just implemented ILM for our indexes.

Here is our ILM Policy :

```
  {
      "policy": {
          "phases": {
              /* Hot phase - Resync Priority 100 */
              "hot": {
                  "min_age": "0ms",
                  "actions": {
                      "set_priority": {
                          "priority": 100
                      }
                  }
              },
              /* Warm phase - Resync Priority 50 */
              "warm": {
                  "min_age": "30d",
                  "actions": {
                      "set_priority": {
                          "priority": 50
                      }
                  }
              },
              /* Cold phase - Resync Priority 0 - Apply storageclass = mass */
              "cold": {
                  "min_age": "90d",
                  "actions": {
                      "allocate": {
                          "require": {
                              "storageclass": "mass"
                          }
                      },
                      "set_priority": {
                          "priority": 0
                      }
                  }
              }
          }
      }
  }
```

Steps Hot and Warm run successfully but we are facing a NullPointer Exception on Cold phase during AllocationRoutedStep

Here is the stacktrace from an index/_ilm/explain :

```
java.lang.NullPointerException
	at org.elasticsearch.xpack.core.indexlifecycle.AllocationRoutedStep.isConditionMet(AllocationRoutedStep.java:70)
	at org.elasticsearch.xpack.indexlifecycle.ExecuteStepsUpdateTask.execute(ExecuteStepsUpdateTask.java:123)
	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:47)
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:643)
	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:270)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:200)
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:135)
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150)
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by DisasteR** [comment]
Really sorry to ask but are there any updates for this ?

**Comment by DisasteR** [comment]
@jimczi do you know if this has a chance to be fixed  #in 6.8 ?

**Comment by joegallo** [comment]
@DisasteR do you have any closed indices? I was able provoke a similar error by creating a policy with an `allocate` step and then closing the index that it applies to, like this:

```
PUT /_cluster/settings
{
  "transient": {
    "indices.lifecycle.poll_interval": "5s"
  }
}

PUT _ilm/policy/policy-63877
{
  "policy": {
    "phases": {
      "cold": {
        "min_age": "30s",
        "actions": {
          "allocate": {
            "number_of_replicas": 2
          }
        }
      }
    }
  }
}

PUT my-index-000001
{
  "settings": {
    "index": {
      "lifecycle": {                                                                                                                                                                                                                                                                                                                                                                "name": "policy-63877"
      }
    }
  }
}

POST my-index-000001/_close

GET _cat/indices?s=index

GET my-index-000001/_ilm/explain
```

**Comment by DisasteR** [comment]
@joegallo yes indices matched by the cold phase are closed.

**Comment by joegallo** [comment]
Okay, if you upgrade to a recent 7.x, then this'll work fine for you (you'd need to re-open and then re-close the relevant indices first, in order for them to be replicated), but on 6.8 this is going to error out. Since it works on 7.x, I don't see a scenario where we're going to create a 6.8-specific fix.

**Comment by DisasteR** [comment]
That's a little rude.. no backport is possible ? I know 6.8 is near eol but maybe it can be done easily ?

I cannot upgrade to 7 for the moment.
So m'y only option is to keep my indices open to get the ilm working ?

**Comment by dakrone** [comment]
@DisasteR in this case there is not a fix that should be backported, because this is fixed in 7.2+ (if it affected our current 7.x versions it might be a different story as far as backporting a fix goes). We also don't consider this a high priority since ILM cannot manage closed indices regardless prior to them being replicated (which happened in 7.2+). For now we should document in 6.8 that ILM cannot manage closed indices (other than deleting them).

**Comment by DisasteR** [comment]
ok thanks for your precisions.
